pkg/lib/import-json.ts

```diff
@@ -0,0 +1,262 @@
+         "players": [
+            { "name": "Haaland", "age": 24 },
+            { "name": "De Bruyne", "age", 33 }
+         ],
+         "stadium": "City of Manchester Stadium"
+       }
+
+    This is a fully valid Team and import_Team will return it without
+    any errors or exceptions. However, the result will not contain the
+    "stadium" field since Team objects don't have that.
+
+       {
+         "name": "ManCity",
+         "players": [
+            { "name": "Haaland", "age": "unknown" },
+            { "name": "De Bruyne", "age", 33 }
+         ]
+       }
+
+    The "age" field of Haaland is not a number, but it is
+    optional. The import_Team function will log an error and will omit
+    the "age" field from the Player object for Håland.
+
+       {
+         "name": "ManCity",
+         "players": [
+            { "name": [ "Erling", "Braut", "Haaland" ], "age": 24 },
+            { "name": "De Bruyne", "age", 33 }
+         ]
+       }
+
+    The "name" field for Håland is not a string, and it is
+    mandatory. An error will be logged and the whole Håland entry is
+    omitted from "players".
+
+      {
+         "name": "ManCity",
+         "players": "TBD"
+       }
+
+    The "players" field is not an array, and it is mandatory. The
+    import_Team function will raise a ValidationError exception.
+ */
+
+/* WRITING IMPORTER FUNCTIONS
+
+   The process of writing a importer function for a given TypeScript
+   interface is pretty mechanic, and could well be automated.
+
+   For example, these are the functions for the Player and Team
+   interfaces from above.
+
+       interface Player {
+          name: string;
+          age?: number;
+       }
+
+       function import_Player(val: JsonValue): Player {
+           const obj = import_json_object(val);
+           const res: Player = {
+               name: import_mandatory(obj, "name", import_string),
+           };
+           import_optional(res, obj, "age", import_number);
+           return res;
+       }
+
+       interface Team {
+          name: string;
+          players: Player[];
+       }
+
+       function import_Team(val: JsonValue): Team {
+           const obj = import_json_object(val);
+           const res: Team = {
+               name: import_mandatory(obj, "name", import_string),
+               players: import_mandatory(obj, "players", v => import_array(v, import_Player))
+           };
+           return res;
+       }
+
+   More examples can be found in "pkg/shell/manifests.ts".
+ */
+
+class ValidationError extends Error { }
+
+const validation_path: string[] = [];
+
+function with_validation_path<T>(p: string, func: () => T): T {
+    validation_path.push(p);
+    try {
+        return func();
+    } finally {
+        validation_path.pop();
+    }
+}
+
+function validation_error(msg: string): never {
+    console.error(`JSON validation error for ${validation_path.join("")}: ${msg}`);
+    throw new ValidationError();
```

Comment:
Couldn't hurt to pass on `msg` into the exception object?

Reply:
Hmm, when being strict, there is only one error, so we can do that.